### ISSUE TITLE
Destroying a promise from within it's callback causes a segfault

### ISSUE THREAD

**Issue by G-glop** [capnproto/capnproto#599]
Minimal example:
Latest master, Windows 10, 64 bit build
```
#include <kj\async-io.h>

int main() {
	auto aio = kj::setupAsyncIo();

	kj::Promise<void> tim = kj::READY_NOW;
	kj::Own<kj::Promise<void> prom = tim.then([&]() {prom = nullptr; });

	prom->wait(aio.waitScope);
}
```

**Comment by kentonv** [comment]
Hi @G-glop,

This is actually expected. Deleting a promise during the continuation is similar to doing "delete this;" in a method. (In fact, in part, this is exactly what's happening: the callback lambda is owned by the Promise, so if you delete the Promise, you are deleting the lambda while it's still executing.)

There's no way to make this work without violating the guarantees of the promise interface. In particular, deleting a Promise is supposed to cancel it. This is supposed to mean that once the destructor completes, the caller has the guarantee that none of the Promise's continuations are running anymore, so objects that those continuations depended on can be deleted. Obviously, if the deletion happens inside the continuation, there's no way to guarantee this.

One hack you can use to work around this is to call `prom.detach()` instead of `prom = nullptr`. This effectively transfers ownership of the promise to the event loop itself, where it will continue to run until completion. Generally, `.detach()` is dangerous because it tends to lead to dangling pointers: often, a Promise holds references to objects it doesn't own, and so you must destroy the Promise before destroying said objects. But, once you've called `.detach()`, you no longer have the ability to destroy the promise. But, if you know that it's the very last thing a promise will do, calling `.detach()` is probably safe.

Still, what I recommend is designing your code so that promises don't delete themselves -- just like you'd design your code so that objects don't `delete this` in the middle of a method call. Generally I have not found it too difficult to do.
